crates/sui-network/src/state_sync.rs

```diff
@@ -0,0 +1,1433 @@
+        }
+
+        info!("State-Synchronizer ended");
+    }
+
+    fn handle_message(&mut self, message: StateSyncMessage) {
+        match message {
+            StateSyncMessage::StartSyncJob => self.maybe_start_sync_job(),
+            StateSyncMessage::VerifiedCheckpoint(checkpoint) => self.handle_checkpoint(checkpoint),
+        }
+    }
+
+    // Handle a checkpoint that we recieved from consensus
+    fn handle_checkpoint(&mut self, checkpoint: Box<VerifiedCheckpoint>) {
+        let (next_sequence_number, previous_digest) = {
+            let latest_checkpoint = self.store.get_latest_checkpoint();
+
+            // If this is an older checkpoint, that's weird but ok, just ignore it
+            if latest_checkpoint.as_ref().map(|x| x.sequence_number())
+                >= Some(checkpoint.sequence_number())
+            {
+                return;
+            }
+
+            let next_sequence_number = latest_checkpoint
+                .as_ref()
+                .map(|x| x.sequence_number().saturating_add(1))
+                .unwrap_or(0);
+            let previous_digest = latest_checkpoint.map(|x| x.digest());
+            (next_sequence_number, previous_digest)
+        };
+
+        // If this is exactly the next checkpoint then lets insert it and gossip it out
+        if checkpoint.sequence_number() == next_sequence_number
+            && checkpoint.previous_digest() == previous_digest
+        {
+            let checkpoint = *checkpoint;
+            self.store.insert_checkpoint(checkpoint.clone());
+
+            // We don't care if no one is listening as this is a broadcast channel
+            let _ = self.checkpoint_event_sender.send(checkpoint.clone());
+
+            self.gossip_checkpoint(checkpoint);
+        } else {
+            // Otherwise shove it in the quarentined checkpoints and we can try to sync the missing
+            // ones
+            self.peer_heights
+                .write()
+                .unwrap()
+                .insert_checkpoint(checkpoint.into_inner());
+            warn!("Consensus gave us too new of a checkpoint");
+        }
+    }
+
+    fn handle_peer_event(
+        &mut self,
+        peer_event: Result<PeerEvent, tokio::sync::broadcast::error::RecvError>,
+    ) {
+        use tokio::sync::broadcast::error::RecvError;
+
+        match peer_event {
+            Ok(PeerEvent::NewPeer(peer_id)) => {
+                self.get_latest_from_peer(peer_id);
+            }
+            Ok(PeerEvent::LostPeer(peer_id, _)) => {
+                self.peer_heights.write().unwrap().heights.remove(&peer_id);
+            }
+
+            Err(RecvError::Closed) => {
+                panic!("PeerEvent channel shouldn't be able to be closed");
+            }
+
+            Err(RecvError::Lagged(_)) => {
+                trace!("State-Sync fell behind processing PeerEvents");
+            }
+        }
+    }
+
+    fn get_latest_from_peer(&mut self, peer_id: PeerId) {
```

Comment:
think it's worth prefixing task-spawning functions with `spawn_` or something like that, to distinguish them?